node/src/blockchain/blockchain_interface/blockchain_interface_web3/lower_level_interface_web3.rs

```diff
@@ -1,395 +1,762 @@
-            .wait()
+impl LowBlockchainInt for LowBlockchainIntWeb3 {
+    fn get_transaction_fee_balance(
+        &self,
+        address: Address,
+    ) -> Box<dyn Future<Item = U256, Error = BlockchainError>> {
+        Box::new(
+            self.web3
+                .eth()
+                .balance(address, None)
+                .map_err(|e| QueryFailed(e.to_string())),
+        )
     }
 
-    fn get_service_fee_balance(&self, wallet: &Wallet) -> ResultForBalance {
-        self.contract
-            .query(
-                "balanceOf",
-                wallet.address(),
-                None,
-                Options::default(),
-                None,
-            )
-            .map_err(|e| BlockchainError::QueryFailed(format!("{} for wallet {}", e, wallet)))
-            .wait()
+    fn get_service_fee_balance(
+        &self,
+        address: Address,
+    ) -> Box<dyn Future<Item = U256, Error = BlockchainError>> {
+        Box::new(
+            self.contract
+                .query("balanceOf", address, None, Options::default(), None)
+                .map_err(|e| QueryFailed(e.to_string())),
+        )
     }
 
-    fn get_block_number(&self) -> LatestBlockNumber {
-        self.web3
-            .eth()
-            .block_number()
-            .map_err(|e| BlockchainError::QueryFailed(e.to_string()))
-            .wait()
+    fn get_gas_price(&self) -> Box<dyn Future<Item = U256, Error = BlockchainError>> {
+        Box::new(
+            self.web3
+                .eth()
+                .gas_price()
+                .map_err(|e| QueryFailed(e.to_string())),
+        )
     }
 
-    fn get_transaction_id(&self, wallet: &Wallet) -> ResultForNonce {
-        self.web3
-            .eth()
-            .transaction_count(wallet.address(), Some(BlockNumber::Pending))
-            .map_err(|e| BlockchainError::QueryFailed(format!("{} for wallet {}", e, wallet)))
-            .wait()
+    fn get_block_number(&self) -> Box<dyn Future<Item = U64, Error = BlockchainError>> {
+        Box::new(
+            self.web3
+                .eth()
+                .block_number()
+                .map_err(|e| QueryFailed(e.to_string())),
+        )
+    }
+
+    fn get_transaction_id(
+        &self,
+        address: Address,
+    ) -> Box<dyn Future<Item = U256, Error = BlockchainError>> {
+        Box::new(
+            self.web3
+                .eth()
+                .transaction_count(address, Some(BlockNumber::Pending))
+                .map_err(move |e| QueryFailed(format!("{} for wallet {}", e, address))),
+        )
+    }
+
+    fn get_transaction_receipt(
+        &self,
+        hash: H256,
+    ) -> Box<dyn Future<Item = Option<TransactionReceipt>, Error = BlockchainError>> {
+        Box::new(
+            self.web3
+                .eth()
+                .transaction_receipt(hash)
+                .map_err(|e| QueryFailed(e.to_string())),
+        )
+    }
+
+    fn get_transaction_receipt_batch(
```

Comment:
**URGENCY STATUS: I WISH YOU DIDN'T IGNORE THIS MISCONCEPTION. I'D REALLY LIKE TO SEE SOMETHING DONE REGARDING THIS. ON THE OTHER HAND, THIS MIGHT BE THE DIRTIEST RECONSTRUCTION AMONG MY PLEAS.** May the God be with you.

Reply:
Points to consider:

1. Make Lower interface unintelligent, such that it's just a wrapper over web3.
2. Find a dividing line between the two interfaces.

- Notes of discussion with the team